xla/service/gpu/fusions/transpose_test.cc

```diff
@@ -246,6 +246,71 @@ TEST_F(TransposeTest, ThreadIndexingPartialBlock) {
       )"));
 }
 
+TEST_F(TransposeTest, ThreadIndexingSidedOutput) {
+  auto module = ParseAndReturnVerifiedModule(R"(
+    HloModule module
+
+    fusion {
+      %input = f32[100,32,64] parameter(0)
+      %bitcast0 = f32[100,64,32] bitcast(%input)
+      %transpose = f32[100,64,32] transpose(%input), dimensions={0,2,1}
+      %bitcast1 = f32[100,2048] bitcast(%bitcast0)
+      ROOT %tuple = (f32[100,64,32], f32[100,2048]) tuple(%transpose, %bitcast1)
+    }
+
+    ENTRY entry {
+      %input = f32[100,32,64] parameter(0)
+      ROOT %fusion = (f32[100,64,32], f32[100,2048]) fusion(%input), kind=kInput, calls=fusion
+    })")
+                    .value();
+
+  auto* root = module->entry_computation()->root_instruction();
+  auto analysis = AnalyzeFusion(*root, device_info_);
+
+  TF_ASSERT_OK_AND_ASSIGN(auto fusion, GetTransposeFusion(analysis));
+  mlir::MLIRContext mlir_context;
+
+  EXPECT_THAT(
+      fusion->ComputeThreadIdToInputIndexing(1, 0, &mlir_context)->ToString(),
```

Comment:
This function is currently a bit ill-defined. It claims to be for the operand of the hero, but that's not a meaningful concept for side outputs.

Your implementation is correct in that it returns the indexing for %bitcast0, but I don't see how that would ever be useful. What we actually need is the indexing for the root itself. I'm not entirely sure how to update the function specification to make that not a mess though.

Can you add a comment here explaining what's going on?

Reply:
Yeah, this function is not used by side outputs. Why I add this test just follow previous test function, maybe I can remove it.

Reply:
I find there exist [usage](https://github.com/openxla/xla/blob/main/xla/service/gpu/model/coalescing_analysis.cc#L167) of this func by side output. So this func has meaning also for side outputs to analysis memory coalescing if side outputs has different input compared with real hero, see this hlo:
fusion {
      %input0 = f32[100,32,64] parameter(0)
      %input1 = f32[100,32,64] parameter(1)
      %bitcast0 = f32[100,64,32] bitcast(%input1)
      %transpose = f32[100,64,32] transpose(%input0), dimensions={0,2,1}
      %bitcast1 = f32[100,2048] bitcast(%bitcast0)
      ROOT %tuple = (f32[100,64,32], f32[100,2048]) tuple(%transpose, %bitcast1)
}
I think we should keep this check.

Reply:
I agree, the check should stay. That's why I asked for a brief explanation in a code comment.